prune/prunables.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2021 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prune
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// Some default pruneable functions
+
+// DefaultPodIsPrunable is a default IsPrunableFunc to be used specifically with Pod resources.
+// This can be overridden by registering your own IsPrunableFunc via the RegisterIsPrunableFunc method
+func DefaultPodIsPrunable(obj client.Object) error {
+	gvk := schema.GroupVersionKind{
+		Group:   "core",
+		Version: "v1",
+		Kind:    "Pod",
+	}
+
+	// if the object is not a Pod then it is not pruneable
+	if obj.GetObjectKind().GroupVersionKind() != gvk {
+		return &ErrUnprunable{
+			Obj:    &obj,
+			Reason: "Object is not a Pod",
+		}
+	}
+
+	// convert to a Pod object so we can do Pod specific checks
+	pod := &corev1.Pod{}
+	err := ConvertClientObjectToKind(obj, pod)
+	if err != nil {
+		return err
+	}
+
+	// If the pod has Succeeded then we can remove it
+	if pod.Status.Phase != corev1.PodSucceeded {
+		return &ErrUnprunable{
+			Obj:    &obj,
+			Reason: "Pod has not succeeded",
+		}
+	}
+
+	return nil
+}
+
+// DefaultJobIsPrunable is a default IsPrunableFunc to be used specifically with Job resources.
+// This can be overridden by registering your own IsPrunableFunc via the RegisterIsPrunableFunc method
+func DefaultJobIsPrunable(obj client.Object) error {
+
+	gvk := schema.GroupVersionKind{
+		Group:   "batch",
+		Version: "v1",
+		Kind:    "Job",
+	}
+
+	if obj.GetObjectKind().GroupVersionKind() != gvk {
+		return &ErrUnprunable{
+			Obj:    &obj,
+			Reason: "Object is not a Job",
+		}
+	}
```

Comment:
Something else to call out about the usage of `ErrUnprunable`. It is not intended to wrap actual errors. It means "this resource should not be pruned due to the business logic of the program."